pkg/kubelet/dockershim/docker_sandbox.go

```diff
@@ -19,33 +19,243 @@ package dockershim
 import (
 	"fmt"
 
+	dockertypes "github.com/docker/engine-api/types"
+	dockercontainer "github.com/docker/engine-api/types/container"
+	dockerfilters "github.com/docker/engine-api/types/filters"
+
 	runtimeApi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
 )
 
+const (
+	defaultSandboxImage = "gcr.io/google_containers/pause-amd64:3.0"
+
+	// Various default sandbox resources requests/limits.
+	defaultSandboxCPUshares int64 = 2
+	defaultSandboxCPUQuota  int64 = 1000
+	// 100000 is equivalent to 100ms
+	defaultSandboxCPUPeriod int64 = 100000
+	defaultSandboxOOMScore  int   = -999
+
+	// Termination grace period
+	defaultSandboxGracePeriod int = 10
+)
+
 // CreatePodSandbox creates a pod-level sandbox.
 // The definition of PodSandbox is at https://github.com/kubernetes/kubernetes/pull/25899
+// For docker, PodSandbox is implemented by a container holding the network
+// namespace for the pod.
+// Note: docker doesn't use LogDirectory (yet).
 func (ds *dockerService) CreatePodSandbox(config *runtimeApi.PodSandboxConfig) (string, error) {
-	return "", fmt.Errorf("Not implemented")
+	// Step 1: Pull the image for the sandbox.
+	// TODO: How should we handle pulling custom pod infra container image
+	// (with credentials)?
+	image := defaultSandboxImage
+	if err := ds.client.PullImage(image, dockertypes.AuthConfig{}, dockertypes.ImagePullOptions{}); err != nil {
+		return "", fmt.Errorf("unable to pull image for the sandbox container: %v", err)
+	}
+
+	// Step 2: Create the sandbox container.
+	createConfig := makeSandboxDockerConfig(config, image)
+	createResp, err := ds.client.CreateContainer(*createConfig)
+	if err != nil || createResp == nil {
+		return "", fmt.Errorf("failed to create a sandbox for pod %q: %v", config.GetName(), err)
+	}
+
+	// Step 3: Start the sandbox container.
+	// Assume kubelet's garbage collector would remove the sandbox later, if
+	// startContainer failed.
+	err = ds.StartContainer(createResp.ID)
+	return createResp.ID, err
 }
 
 // StopPodSandbox stops the sandbox. If there are any running containers in the
 // sandbox, they should be force terminated.
 func (ds *dockerService) StopPodSandbox(podSandboxID string) error {
-	return fmt.Errorf("Not implemented")
+	return ds.client.StopContainer(podSandboxID, defaultSandboxGracePeriod)
```

Comment:
@yujuhong Yeah, I think the code added won't be used in most cases. But we've explicitly defined the behaviour in runtime interface, we may want to conform with it. :)